### PR TITLE
Async RL prototype

### PR DESCRIPTION
**Pull request by Darktex**
#### Context
What is the purpose of this PR? Is it to
- [X] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

This PR adds a prototype for async RL in TorchTune. Compared to the sync version of this code, we are seeing over 10x speedup. The API still has a few rough edges, and we are looking to make this more generic for the future, but this represents a good first step towards a fully async RL solution.

#### Changelog
What are the changes made in this PR? TLDR is:

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [X] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [X] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [X] run unit tests via `pytest tests`
- [X] run recipe tests via `pytest tests -m integration_test`
- [X] manually run any new or modified recipes with sufficient proof of correctness
- [X] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: torchtune/modules/transformer.py [hunk 1]
```diff
@@ -576,6 +576,7 @@ def forward(
         encoder_mask: Optional[torch.Tensor] = None,
         input_pos: Optional[torch.Tensor] = None,
         input_embeds: Optional[torch.Tensor] = None,
+        output_mask: Optional[torch.Tensor] = None,
```
Comment (Darktex) [inline torchtune/modules/transformer.py:579]: Comment from @felipemello1 is to not do this and have this done in the loss instead. @pbontrager @ebsmothers @joecummings

File: pyproject.toml [hunk 2]
```diff
@@ -35,6 +35,7 @@ dependencies = [
 
     # Multimodal
     "Pillow>=9.4.0",
+    "math-verify[antlr4_9_3]"
```
Comment (joecummings) [inline pyproject.toml:38]: Can you put this in the ``async_rl`` requirements rather than the main requirements?

File: recipes/dev/async_grpo.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+git clone https://github.com/joecummings/r1-zero.git
+cd r1-zero
+pip install torch torchvision torchao
+pip install -e .[async_rl]
+
+```
+
+With these installed you can run the recipe with
+
+```bash
```
Comment (joecummings) [inline recipes/dev/async_grpo.md:62]: Based on Davide I's comments, we should probably explicitly instruct people to download the model in this guide.

File: recipes/dev/async_grpo.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+tune run dev/async_grpo_full_finetune_distributed --config recipes/configs/dev/qwen3B_async_grpo.yaml
```
Comment (joecummings) [inline recipes/dev/async_grpo.md:63]: We don't actually need the YAML ending. We should refer to it by the way we do other things e.g. `dev/qwen3B_async_grpo`

File: torchtune/dev/rl/rewards.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+            reward, success = shaped_correctness_reward(
+                answer=answers[b], completion=text_completion
+            )
+            rewards[b, g] = reward
+            successes[b, g] = success
+
+    return rewards, successes
+
+
+# class ShapedCorrectnessReward(Transform):
```
Comment (joecummings) [inline torchtune/dev/rl/rewards.py:161]: Can this be deleted?

**Review by joecummings — COMMENTED** [review]
Few nits

File: recipes/configs/dev/3B_full_grpo.yaml [hunk 6]
```diff
@@ -3,7 +3,7 @@
 #
 # This config assumes that you've run the following command before launching
 # this run:
-#   tune download meta-llama/Llama-3.2-3B --output-dir /tmp/Llama-3.2-3B --ignore-patterns "original/consolidated.00.pth"
+#   tune download Qwen/Qwen2.5-3B --output-dir /tmp/Qwen2.5-3B
```
Comment (ebsmothers) [inline recipes/configs/dev/3B_full_grpo.yaml:6]: This shouldn't be changed, right? Isn't this still the sync version using Llama 3.2 3B?

File: recipes/configs/dev/qwen3B_async_grpo.yaml [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+# To launch, run the following:
+# tune run dev/async_grpo --config recipes/configs/dev/qwen3B_async_grpo.yaml
+
+# Note that unlike in synchronous recipes (all the existing ones), we don't follow a SPMD model here so passing a flag like `--nproc-per-node 8`
+# won't help here. This is instead launching just the controller, which is a CPU process. The controller will in turn launch every other worker.
+
+name: grpo_async_qwen3b
+
+output_dir: /tmp/checkpoints/${name}
+base_model_path: /tmp/Qwen2.5-3B  # Use this to train from the slightly trained SFT model
```
Comment (ebsmothers) [inline recipes/configs/dev/qwen3B_async_grpo.yaml:21]: nit
```suggestion
base_model_path: /tmp/Qwen2.5-3B  
```

File: recipes/configs/dev/qwen3B_sync_grpo.yaml [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+# using a Qwen-3B Base model (which is known to train quickly on this task).
+#
+# Also check out the async implementation of this workflow in this directory.
+#
+# This config assumes that you've run the following command before launching
+# this run:
+#   tune download Qwen/Qwen2.5-3B --output-dir /tmp/Qwen2.5-3B
+#
+# To launch on 4 devices, run the following command from root:
+# tune run --nproc_per_node 8 dev/grpo_full_finetune_distributed --config dev/qwen3B_sync_grpo
```
Comment (ebsmothers) [inline recipes/configs/dev/qwen3B_sync_grpo.yaml:11]:
```suggestion
# tune run --nproc_per_node 4 dev/grpo_full_finetune_distributed --config dev/qwen3B_sync_grpo
```

File: recipes/configs/dev/qwen3B_sync_grpo.yaml [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+# to override the checkpointer directory while launching training
+# you can run:
+#   tune run --nproc_per_node 4 dev/grpo_full_finetune_distributed --config dev/qwen3B_sync_grpo checkpointer.checkpoint_dir=<YOUR_CHECKPOINT_DIR>
+#
+# This config works best when the model is being fine-tuned on 2+ GPUs.
+#
+# Furthermore, you can launch it on multiple nodes by going to recipes/dev/ and using
+#  sbatch multinode_grpo.sbatch
+
+name: grpo_async_qwen3b
```
Comment (ebsmothers) [inline recipes/configs/dev/qwen3B_sync_grpo.yaml:23]:
```suggestion
name: grpo_sync_qwen3b
```

File: recipes/configs/dev/qwen3B_async_grpo.yaml [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+# This config assumes that you've run the following command before launching
+# this run:
+#   tune download Qwen/Qwen2.5-3B --output-dir /tmp/Qwen2.5-3B --ignore-patterns "original/consolidated.00.pth"
+#
+# This config is meant to run on a whole node with 8 GPUs. You can tweak the number of workers here to change it
+# but note that some workers have currently been hardcoded in the prototype and we haven't yet tested on different
+# numbers of workers.
+
+# To launch, run the following:
+# tune run dev/async_grpo --config recipes/configs/dev/qwen3B_async_grpo.yaml
```
Comment (ebsmothers) [inline recipes/configs/dev/qwen3B_async_grpo.yaml:13]: Is this right? I thought the recipe name was `dev/async_grpo_full_finetune_distributed`

File: torchtune/dev/rl/generation.py [hunk 11]
```diff
@@ -0,0 +1,244 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (ebsmothers) [inline torchtune/dev/rl/generation.py:1]: How much of generation.py, gsm8k.py, data.py, rewards.py, types.py is identical to what's already in dev/grpo? (At least for this file they seem identical.) Obv not for this PR but we should consolidate

File: .github/workflows/regression_test.yaml [hunk 12]
```diff
@@ -57,8 +57,9 @@ jobs:
         run: python -m pip install torch torchvision torchao
       - name: Install remaining dependencies
         run: |
-          python -m pip install -e ".[dev]"
           python -m pip install lm-eval>=0.4.5
+          python -m pip install -e ".[dev]"
+          python -m pip install -e ".[async_rl]"
```
Comment (ebsmothers) [inline .github/workflows/regression_test.yaml:62]: Do we actually need this? I thought our regression test is basically a no-op at this point

File: .github/workflows/gpu_test.yaml [hunk 13]
```diff
@@ -52,8 +52,10 @@ jobs:
         run: python -m pip install torch torchvision torchao
       - name: Install recipe-specific dependencies
         run: python -m pip install lm-eval>=0.4.5
-      - name: Install the torchtune library with dev option!
-        run: python -m pip install -e ".[dev]"
+      - name: Install the torchtune library with dev and async_rl options!
+        run: |
+          python -m pip install -e ".[dev]"
+          python -m pip install -e ".[async_rl]"
```
Comment (ebsmothers) [inline .github/workflows/gpu_test.yaml:58]: Weren't we gonna use a separate workflow for the async RL tests?

File: torchtune/dev/grpo/rewards.py [hunk 14]
```diff
@@ -94,3 +94,83 @@ def batch_shaped_correctness_reward(
             successes[b, g] = success
 
     return rewards, successes
+
+
+# class ShapedCorrectnessReward(Transform):
```
Comment (ebsmothers) [inline torchtune/dev/grpo/rewards.py:99]: bump the previous comment to remove this (seems to have disappeared)

File: torchtune/dev/rl/rewards.py [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+from typing import List, Tuple
+from xml.etree import ElementTree as ET
+
+import math_verify
+
+import torch
+
+# from tensordict import TensorDict, TensorDictBase
+# from torchrl.data import Composite, TensorSpec, Unbounded
+# from torchrl.envs import Transform
```
Comment (ebsmothers) [inline torchtune/dev/rl/rewards.py:17]: remove

File: torchtune/dev/rl/workers/postprocessing.py [hunk 16]
```diff
@@ -0,0 +1,339 @@
...
+            }
+        )
+
+        # # TODO we should encode this in the dataclass instead of keeping a dict
+        # # otherwise we end up with a list of identical dicts
+        # assert all(
+        #     metadata["func_names"] == reward_metadata[0]["func_names"]
+        #     for metadata in reward_metadata
+        # ), "Function names in reward_metadata are not consistent across all entries"
+        # function_names = reward_metadata[0]["func_names"]
```
Comment (ebsmothers) [inline torchtune/dev/rl/workers/postprocessing.py:166]: remove

File: torchtune/dev/rl/workers/postprocessing.py [hunk 17]
```diff
@@ -0,0 +1,339 @@
...
+        )
+
+        log_dict.update(
+            {
+                "ref_actor_performance/time_total_ref_step (s)": time_total_ref_step,
+                "ref_actor_performance/time_model_running (s)": time_model_running,
+                "ref_actor_performance/pct_time_model_running (%)": pct_time_model_running,
+                "ref_actor_performance/time_waiting_buffer (s)": time_waiting_buffer,
+                "ref_actor_performance/pct_time_waiting_buffer (%)": pct_time_waiting_buffer,
+                # "queues/ref_actor_full_queue_data_discard": full_queue_data_discard,
```
Comment (ebsmothers) [inline torchtune/dev/rl/workers/postprocessing.py:148]: remove

File: torchtune/modules/transformer.py [hunk 18]
```diff
@@ -695,14 +699,18 @@ def forward(
...
-        output = self.unembed(h)
+        output = self.unembed(h, output_mask=output_mask)
 
         # Output list if hidden states are requested, otherwise just the output
         # TODO: always output a list to have a consistent output type
         output = output if not hidden else [*hidden, output]
         return output
 
-    def unembed(self, h):
+    def unembed(self, h, output_mask=None):
```
Comment (ebsmothers) [inline torchtune/modules/transformer.py:709]: need to revert these changes

File: recipes/dev/async_grpo.md (hunk 3 above)
Comment (Darktex) [reply recipes/dev/async_grpo.md:69]: Done!

File: recipes/dev/async_grpo.md (hunk 4 above)
Comment (Darktex) [reply recipes/dev/async_grpo.md:70]: 🤯

File: torchtune/dev/rl/generation.py (hunk 11 above)
Comment (Darktex) [reply torchtune/dev/rl/generation.py:1]: Rewards is quite different, the rest are closer. We can def consolidate down the road...

File: recipes/configs/dev/qwen3B_async_grpo.yaml (hunk 10 above)
Comment (Darktex) [reply recipes/configs/dev/qwen3B_async_grpo.yaml]: Great catch thanks

File: pyproject.toml (hunk 2 above)
Comment (Darktex) [reply pyproject.toml]: Done!

File: torchtune/dev/grpo/rewards.py (hunk 14 above)
Comment (SalmanMohammadi) [reply torchtune/dev/grpo/rewards.py:99]: Yeah this should be removed. This has also been refactored in https://github.com/joecummings/r1-zero/pull/48 which I'll follow up this PR with

**Review by ebsmothers — APPROVED** [review]

**Review by ebsmothers — APPROVED** [review]
